### ISSUE TITLE
`VertexAIEnvironment` to support multi-node training on Vertex AI clusters

### ISSUE THREAD

**Issue by ordabayevy** [cellarium-ai/scvi-distributed#56]
Add [`ClusterEnvironment`](https://lightning.ai/docs/pytorch/stable/api_references.html#environments) to support distributed training on Vertex AI.

In particular, it seems `os.environ["WORLD_SIZE"]` variable is set to number of nodes and needs to be adjusted to total number of GPUs. Node rank is store in `os.environ["RANK"]` variable.

**Comment by agrawalraj** [comment]
I can take this issue! This looks promising: https://devblog.pytorchlightning.ai/how-to-configure-a-gpu-cluster-to-scale-with-pytorch-lightning-part-2-cf69273dde7b

**Comment by ordabayevy** [comment]
Thanks @agrawalraj ! 

I can share some tips from my past experience of working with Vertex AI:

### Submitting a job

Training jobs can be submitted to Vertex AI using a python script. More about it - https://cloud.google.com/vertex-ai/docs/training/overview

Here is the script that I was using to submit a custom job to Vertex AI:

<details>

```py
# Copyright Contributors to the Cellarium project.
# SPDX-License-Identifier: BSD-3-Clause

from google.cloud import aiplatform


def create_custom_job(
    project: str,
    location: str,
    staging_bucket: str,
    display_name: str,
    script_path: str,
    config_file: str,
    container_uri: str,
    replica_count: int = 1,
    machine_type: str = "n1-highmem-16",
    accelerator_type: str = "NVIDIA_TESLA_T4",
    accelerator_count: int = 4,
    boot_disk_type: str = "pd-ssd",
    boot_disk_size_gb: int = 100,
    labels: dict[str, str] | None = None,
    experiment: str | None = None,
    experiment_run: str | None = None,
) -> None:
    aiplatform.init(project=project, location=location, staging_bucket=staging_bucket)

    args = ["fit", "--config", config_file]

    worker_pool_specs = aiplatform.utils.worker_spec_utils._DistributedTrainingSpec.chief_worker_pool(
        replica_count=replica_count,
        machine_type=machine_type,
        accelerator_count=accelerator_count,
        accelerator_type=accelerator_type,
        boot_disk_type=boot_disk_type,
        boot_disk_size_gb=boot_disk_size_gb,
    ).pool_specs

    for spec_order, spec in enumerate(worker_pool_specs):
        if not spec:
            continue

        command = [
            "sh",
            "-c",
            # f"python3 {script_path}",
            f"python3 {script_path} {' '.join(args)}",
        ]

        spec["container_spec"] = {
            "image_uri": container_uri,
            "command": command,
        }

    job = aiplatform.CustomJob(
        display_name=display_name,
        worker_pool_specs=worker_pool_specs,
        # base_output_dir=base_output_dir,
        project=project,
        location=location,
        # credentials=credentials,
        labels=labels,
        # encryption_spec_key_name=encryption_spec_key_name,
        staging_bucket=staging_bucket,
    )

    job.run(
        # experiment=experiment,
        # experiment_run=experiment_run,
    )


if __name__ == "__main__":
    create_custom_job(
        project="dsp-cell-annotation-service",
        location="us-central1",
        staging_bucket="gs://dsp-cell-annotation-service",
        #  display_name="onepass324",
        #  script_path="/gcs/dsp-cell-annotation-service/scvid/config_files/onepass_mean_var_std.py",
        #  config_file="/gcs/dsp-cell-annotation-service/scvid/config_files/onepass_mean_var_std.yaml",
        display_name="ppca324node9gpu4accum9",
        script_path="/gcs/dsp-cell-annotation-service/scvid/config_files/probabilistic_pca.py",
        config_file="/gcs/dsp-cell-annotation-service/scvid/config_files/probabilistic_pca.yaml",
        #  display_name="node3gpu4accum1",
        #  script_path="/gcs/dsp-cell-annotation-service/scvid/config_files/grad.py",
        #  config_file="/gcs/dsp-cell-annotation-service/scvid/config_files/probabilistic_pca.yaml",
        # container_uri="us-docker.pkg.dev/vertex-ai/training/pytorch-gpu.1-13.py310:latest",
        container_uri="us-central1-docker.pkg.dev/dsp-cell-annotation-service/scvid-docker-repo/scvid:lightning-cli",
        # container_uri="us-docker.pkg.dev/vertex-ai/training/pytorch-gpu.1-13.py310:latest",
        # container_uri="pytorch/pytorch:latest",
        replica_count=9,
        accelerator_count=4,
        #  labels: dict[str, str] | None = None,
        #  experiment="onepass",
        #  experiment_run: str | None = None,
    )
```

</details>

### Docker image

Running on Vertex AI requires to have a Docker image of `cellarium-ml` saved as a GCP artifact. Above it's given as a `container_uri` argument. I based my Docker image on a pre-built [PyTorch 1.13 container](https://cloud.google.com/vertex-ai/docs/training/pre-built-containers#pytorch), however, since then we moved on to PyTroch 2 and deprecated PyTorch 1.13. So unfortunately, at the moment we can't use Vertex AI's pre-built containers. Few other considerations I have in mind:

1. `cellarium-ml` requirements: PyTorch 2 and Python 3.10
2. In the above custom job script I provide both `script_file` path and `config_file` path. This gave me flexibility to have one docker image and play around with the script file and the config file. It makes total sense to keep config file path as an argument because it will always change. As for the script file it presumably can be hard coded to only the list of available scripts from the `examples` folder but not sure if it provides any benefits.
3. Installing `cellarium-ml`. As far as I know the package can be installed by copying the source code and installing locally or pip installed from the github repo (I used the latter). Not sure if any of these two has an advantage over the other.
4. Automate dockerization with GitHub actions. We most likely want to dockerize every released version of `cellarium-ml`. Do we need to dockerize every time we merge to the main branch? There is a separate issue that's already open about automating dockerization - https://github.com/cellarium-ai/cellarium-ml/issues/58

### ClusterEnvironment

We need to create a `VertexAIEnvironment` as shown in the @agrawalraj 's tutorial link above. The default [LightningEnvironment](https://lightning.ai/docs/pytorch/stable/_modules/lightning/fabric/plugins/environments/lightning.html#LightningEnvironment) in PyTorch Lightning can be used as a template. If I remember correctly `os.environ["WORLD_SIZE"]` corresponded to the number of nodes and the `os.environ["RANK"]` corresponded to the node rank. So for example the world size should be equal to:

```py
def world_size(self):
    return `os.environ["WORLD_SIZE"]` * num_gpus
```

I couldn't find a documentation about Vertex AI environment variables. One way to inspect them is to just print `os.environ` and see what do they match with.

### Testing

It would be nice to have some tests that check that the world size and ranks make sense. We can run the tests locally and mark to skip them on Github actions.

**Comment by agrawalraj** [comment]
@ordabayevy thanks for the very clear and detailed response! When I trained `geneformer` on GCP I ended up using the Deep Learning VM image which uses PyTorch 2. I think we can use this image with Vertex AI according to this article: https://cloud.google.com/vertex-ai/docs/general/deep-learning. The nice thing with the Deep Learning VM image is that it stays up to date with new PyTorch releases.

**Comment by ordabayevy** [comment]
That's interesting. Do you know if we can use those DL VMs when spinning up multiple VMs on Vertex AI? If yes, that would be cool.

**Comment by ordabayevy** [comment]
From the `pytorch-lightning` slack conversation:

**Yerdos Ordabayev:**

 Hi all,
When I'm doing a custom model training on Vertex AI with multiple nodes PyTorch Lightning's ddp strategy chooses `KubeflowEnvironment` as a `cluster_environment`. This sets `world_size=os.environ["WORLD_SIZE"]` which is equal to the number of nodes. However, I believe `world_size` should be `num_nodes*devices`. Is  `KubeflowEnvironment`  is the right choice for multi node (and multi-gpu on each node) training on Vertex AI VMs? If yes, how can I fix the `world_size`? Thank you!

**Adrian:**

How are you launching your multi-node job on Vertex? Sorry, haven't used this platform. world size is normally the number of total processes, so the assignment `world_size=os.environ["WORLD_SIZE"]` is correct and the cluster should give the right value there. If the automatic detection is getting in the way, you can force the default cluster environment like so:

```py
from lightning.pytorch.plugins.environments import LightningEnvironment

trainer = Trainer(..., plugins=LightningEnvironment())
```
This will force the Trainer not to detect Kubeflow and just use regular distributed setup. (edited) 


**Yerdos Ordabayev:**

Thank you for your reply. There is not a lot of information about env variables for Vertex AI clusters but I have found this Vertex AI tutorial for PyTorch ([https://colab.research.google.com/github/GoogleCloudPlatform/vertex-ai-samples/blob/main/[…]rver/pytorch_distributed_training_reduction_server.ipynb](https://colab.research.google.com/github/GoogleCloudPlatform/vertex-ai-samples/blob/main/notebooks/official/reduction_server/pytorch_distributed_training_reduction_server.ipynb)) where `os.environ["WORLD_SIZE"]` is adjusted from being a total number of nodes to a total number of processes:

```py
ngpus_per_node = torch.cuda.device_count()
world_size = int(os.environ["WORLD_SIZE"])

# Since we have ngpus_per_node processes per node, the total world_size
# needs to be adjusted accordingly
world_size =  world_size * ngpus_per_node
```

and global rank is computed from the node rank (os.environ["RANK"]):

```py
# This is the (global) rank of the current process
rank = int(os.environ["RANK"])

# For multiprocessing distributed training, rank needs to be the
# global rank among all the processes
rank = rank * ngpus_per_node + local_rank
```

I followed the tutorial above and [RUN ON AN ON-PREM CLUSTER (INTERMEDIATE)](https://lightning.ai/docs/pytorch/latest/clouds/cluster_intermediate_1.html)  tutorial to adjust `os.environ["WORLD_SIZE"]` and `os.environ["NODE_RANK"]` in my script and I also forced `LightningEnvironment` as you suggested so that `KubeflowEnvironment` won't get detected automatically. This seems to work but I'm not sure if this is the right way to do it since it feels kinda hacky (I was expecting it would just work out of the box).

1. If this is an interest I would love to contribute `VertexAIEnvironment` (with guidance from lightning developers) so that PyTorch Lightning would work on Vertex AI clusters out of the box.
2. I'm still trying to understand how multi-node training works with Lightning and it's not clear to me how it interoperates `torch.distributed.launch` and `torchrun`. My understanding is that unless you [explicitly call](https://lightning.ai/docs/pytorch/latest/clouds/cluster_intermediate_2.html) torch.distributed.run PyTorch Lightning uses it's own way of spawning processes, right?
(edited)

colab.research.google.comcolab.research.google.com
[Google Colaboratory](https://colab.research.google.com/github/GoogleCloudPlatform/vertex-ai-samples/blob/main/notebooks/official/reduction_server/pytorch_distributed_training_reduction_server.ipynb) (5 kB)
https://colab.research.google.com/github/GoogleCloudPlatform/vertex-ai-samples/blob/main/notebooks/official/reduction_server/pytorch_distributed_training_reduction_server.ipynb


**Yerdos Ordabayev:**

 I'm launching it using Vertex AI's [CustomJob](https://cloud.google.com/python/docs/reference/aiplatform/latest/google.cloud.aiplatform.CustomJob) API where the command that I provide to each node (container) is:

```
python train.py fit --config config.yaml
```

where `train.py` looks like this:

```py
import os
from lightning.pytorch.cli import LightningCLI
import torch

from scvid.data import DistributedAnnDataCollectionDataModule
from scvid.train import TrainingPlan


class PPCALightningCLI(LightningCLI):
    """LightningCLI with custom argument linking."""

    def add_arguments_to_parser(self, parser):
        parser.link_arguments(
            "data.n_obs", "model.module.init_args.n_cells", apply_on="instantiate"
        )
        parser.link_arguments(
            "data.n_vars", "model.module.init_args.g_genes", apply_on="instantiate"
        )


def main():
    os.environ["NODE_RANK"] = os.environ.get("RANK")
    ngpus_per_node = torch.cuda.device_count()
    world_size = int(os.environ["WORLD_SIZE"])
    os.environ["WORLD_SIZE"] = str(world_size * ngpus_per_node)
    PPCALightningCLI(
        TrainingPlan,
        DistributedAnnDataCollectionDataModule,
    )


if __name__ == "__main__":
    main()
```


**Adrian:**

> 1. If this is an interest I would love to contribute VertexAIEnvironment (with guidance from lightning developers) so that PyTorch Lightning would work on Vertex AI clusters out of the box.

In general yes, contributions welcome, if it is useful for the community. Opening a GH issue would be a first start, but you could develop this also just for yourself to get unblocked. I would start by making a dummy job and just print os.environ to check all the variables that are being set. Then we can make sense of it and know what to parse for Lightning. It's probably also documented on their side. I don't really have much capacity to help hands-on, but I can help answer questions for sure.

> 2. I'm still trying to understand how multi-node training works with Lightning and it's not clear to me how it interoperates torch.distributed.launch and torchrun. My understanding is that unless you [explicitly call](https://lightning.ai/docs/pytorch/latest/clouds/cluster_intermediate_2.html) torch.distributed.run PyTorch Lightning uses it's own way of spawning processes, right?

Yes, you understand right. What Lightning does is very similar, if not indistinguishable from torchrun so pretty straightforward.

**Yerdos Ordabayev:**

Thank you so much. It's much clearer now.

**Comment by ordabayevy** [comment]
https://github.com/Lightning-AI/pytorch-lightning/pull/19911
